🚨 Low code CDK: Add session token authenticator

flash1293 opened the pull request:
## What

Fixes https://github.com/airbytehq/airbyte/issues/26341

This PR adds a new authenticator: The `SessionTokenAuthenticator`. The existing authenticator under the same name is renamed to `LegacySessionTokenAuthenticator`. The metabase connector is affected by this change, I will update it in a separate PR.

As part of the implementation, the `Requester` interface is extended with a `send_request` method. The intercom connector is affected by this, I will update in a separate PR.

```yaml
  SessionTokenAuthenticator:
    type: object
    required:
      - type
      - login_requester
      - session_token_path
      - request_authentication
    properties:
      type:
        type: string
        enum: [GenericSessionTokenAuthenticator]
      login_requester:
        title: Login Requester
        description: Description of the request to perform to obtain a session token to perform data requests. The response body is expected to be a JSON object with a session token property.
        "$ref": "#/definitions/HttpRequester"
        examples:
          - type: HttpRequester
            url_base: "https://my_api.com"
            path: "/login"
            authenticator:
              type: BasicHttpAuthenticator
              username: "{{ config.username }}"
              password: "{{ config.password }}"
      session_token_path:
        title: Session Token Path
        description: The path in the response body returned from the login requester to the session token.
        examples:
          - ["access_token"]
          - ["result", "token"]
        type: array
        items:
          type: string
      expiration_time:
        title: Expiration Time
        description: The duration in ISO 8601 duration notation after which the session token expires, starting from the time it was obtained. Omitting it will result in the session token being refreshed for every request.
        type: string
        examples:
          - "PT1H"
          - "P1D"
      request_authentication:
        title: Data request authentication
        description: Authentication method to use for requests sent to the API, specifying how to inject the session token.
        anyOf:
          - "$ref": "#/definitions/SessionTokenRequestApiKeyAuthenticator"
          - "$ref": "#/definitions/SessionTokenRequestBearerAuthenticator"
      $parameters:
        type: object
        additionalProperties: true
  SessionTokenRequestApiKeyAuthenticator:
    type: object
    title: API Key Authenticator
    description: Authenticator for requests using the session token as an API key that's injected into the request.
    required:
      - type
      - inject_into
    properties:
      type:
        enum: [ApiKey]
      inject_into:
        title: Inject API Key Into Outgoing HTTP Request
        description: Configure how the API Key will be sent in requests to the source API.
        "$ref": "#/definitions/RequestOption"
        examples:
          - inject_into: header
            field_name: Authorization
          - inject_into: request_parameter
            field_name: authKey
  SessionTokenRequestBearerAuthenticator:
    title: Bearer Authenticator
    description: Authenticator for requests using the session token as a standard bearer token.
    required:
      - type
    properties:
      type:
        enum: [Bearer]
```

## Behavior

* Configure a regular requester for the login request as it allows to reuse the logic in place (error handling / authentication mechanisms)
* Allow session token extraction from the response via path
* Static expiry times (ISO duration)
* request_authentication is specifying the authentication scheme for the actual data requests
   * SessionTokenAuthenticator uses the existing apikey and bearer authenticator classes internally to implement the logic
   * It’s not using the existing component in the yaml directly because the token is controlled by the session token authenticator instead of passed in from the config

The authenticator instance holds a reference to the resolved token. If there is none, execute the login request as defined in the configuration. If it returns an OK status code, it will execute the session token interpolated string. If it fails, it raises. If it returns something, it will assume it’s authenticated and execute the defined request_authentication on data requests.

The expiry time defines how long it will use the cached token before attempting to re-authenticate.

## Example

The metabase API requires an authentication scheme that matches the behavior described above:
https://www.metabase.com/learn/administration/metabase-api#authenticate-your-requests-with-a-session-token

It can be configured the following way:
```yaml
    authenticator:
      type: SessionTokenAuthenticator
      login_requester:
        type: HttpRequester
        url_base: "{{ config['instance_api_url'] }}"
        path: "/session"
        http_method: POST
        request_body_json:
          username: "{{ config.username }}"
          password: "{{ config.password }}"
      expiration_time: P10D
      session_token_path: ["id"]
      request_authentication:
        type: ApiKey
        inject_into:
          type: RequestOption
          inject_into: header
          field_name: X-Metabase-Session
```

## How


### `requesters/requester.py`

Extend the declarative requester interface with a `send_request` method to "execute" the requester and actually perform the request.

### `requesters/http_requester.py`

Pull in existing logic for preparing the request, applying the error handler and the authenticator from the `SimpleRetriever` and `HttpStream` classes to implement the `send_request` interface. In a separate step, the simple retriever will be switched over to use this logic as well.

### `auth/token.py` and `auth/token_provider.py`

To be able to reuse existing implementations of API key and bearer authenticator in the context of the session token authenticator, this PR splits out the concept of token retrieval from the authenticators itself into a `TokenProvider`. This token provider is implemented in two ways:
* The `InterpolatedStringTokenProvider` which encapsulates the existing logic of executing the jinja expression with the current configuration (used in all existing cases)
* The `SessionTokenProvider` which implements the logic to do an auxiliary request to obtain the session token which will be cached for the configured amount of time before a new token is requested. This new provider is leveraging the new `send_request` method defined on the requester.

### `parsers/model_to_component_factory.py`

Extend the ModelToComponentFactory to handle session token authenticators. The session token authenticator defined in the yaml is mapped to the token provider concept by creating a session token provider instance and either an ApiKeyAuthenticator or a BearerAuthenticator based on the value of the `request_authentication` property.


## 🚨 User Impact 🚨

**The existing `SessionTokenAuthenticator` in the low code framework is renamed to `LegacySessionTokenAuthenticator`.** For existing connectors using `SessionTokenAuthenticator`, the `type` property of the authenticator definition in the yaml has to be switched when upgrading.

Old
```
    authenticator:
      type: "SessionTokenAuthenticator"
      username: "{{ config['username'] }}"
      password: "{{ config['password'] }}"
      # ...
```

New
```
    authenticator:
      type: "LegacySessionTokenAuthenticator"
      username: "{{ config['username'] }}"
      password: "{{ config['password'] }}"
      # ...
```

**The `Requester` interface is extended with a `send_request` method.** If `CustomRequester` is used, it's required the provided class implements `send_request` in a way that sends out the request (with error handler, request options and authenticator options applied) and returns the response.

**The `use_cache` and `cache_filename` properties are removed from the  `Requester` interface as they were unused before. It's required to adjust code that interacts with generic Requester objects to not access these properties anymore.

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py:
```diff
@@ -138,6 +138,21 @@ def request_kwargs(
+    @abstractmethod
+    def send_request(
+        self,
+        path: Optional[str] = None,
```
when would the path be None?

flash1293 replied:
In most cases - it's only not none if the caller wants to override the "regular" path which happens for path injection in the paginator

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py:
```diff
@@ -155,3 +168,343 @@ def cache_filename(self) -> str:
     def use_cache(self) -> bool:
         # FIXME: this should be declarative
```
nit: can you update this comment? request caching isn't possible with this requester because [it doesn't use request_cache like HttpStream does](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py#L41)

girarda replied:
otherwise I'd also be fine removing the methods from the interface

flash1293 replied:
I will remove it for now

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py:
```diff
@@ -155,3 +168,343 @@ def cache_filename(self) -> str:
+    def max_retries(self) -> Union[int, None]:
+        if self.disable_retries:
+            return 0
+        if hasattr(self.error_handler, "max_retries"):
```
I would expect `max_retries` to always be set since it is part of the ErrorHandler interface

flash1293 replied:
I can return None though - I will adjust

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/auth/token_provider.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@dataclass
+class SessionTokenProvider(TokenProvider):
```
Q: should we eventually update the OAuth authentication to also use a token provider? It looks like a similar pattern as the session token authenticator to me

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/auth/token.py:
```diff
@@ -172,7 +169,7 @@ def get_new_session_token(api_url: str, username: str, password: str, response_k
 
 @dataclass
-class SessionTokenAuthenticator(DeclarativeAuthenticator):
+class LegacySessionTokenAuthenticator(DeclarativeAuthenticator):
```
what is the trigger for deleting the legacy implementation?

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:
```diff
@@ -252,16 +260,49 @@ def create_api_key_authenticator(model: ApiKeyAuthenticatorModel, config: Config
     @staticmethod
-    def create_bearer_authenticator(model: BearerAuthenticatorModel, config: Config, **kwargs) -> BearerAuthenticator:
+    def create_bearer_authenticator(
+        model: BearerAuthenticatorModel, config: Config, token_provider: Optional[TokenProvider] = None, **kwargs
```
I think we should update the interface to pass in a `Union[SessionTokenAuthenticatorModel, TokenProvider]`

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:
```diff
@@ -252,16 +260,49 @@ def create_api_key_authenticator(model: ApiKeyAuthenticatorModel, config: Config
+    def create_session_token_authenticator(
+        self, model: SessionTokenAuthenticatorModel, config: Config, **kwargs
+    ) -> Union[ApiKeyAuthenticator, BearerAuthenticator]:
+        login_requester = self._create_component_from_model(model=model.login_requester, config=config, name="login_requester")
```
why does it need a name?

girarda replied:
@flash1293 is the name field used for anything?

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:
```diff
@@ -1010,6 +1010,81 @@ definitions:
+        description: Description of the request to perform to obtain a session token to perform data requests. The response body is expected to be a JSON object with a session token property.
+        "$ref": "#/definitions/HttpRequester"
+        examples:
+          - type: HttpRequester
```
I wonder how long it'll take before we start seeing nested SessionTokenAuthenticators...

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:
```diff
@@ -232,7 +238,9 @@ def create_add_fields(self, model: AddFieldsModel, config: Config, **kwargs) ->
 
     @staticmethod
-    def create_api_key_authenticator(model: ApiKeyAuthenticatorModel, config: Config, **kwargs) -> ApiKeyAuthenticator:
+    def create_api_key_authenticator(
```
I think we should update the interface to pass in a Union[SessionTokenAuthenticatorModel, TokenProvider]

With the current implementation, it's not obvious that the model is ignored if a token provider is set. We could add a check to verify that the model.api_token is `""`,  but using a `Union` makes it clear that we only expect one or the other

flash1293 replied:
The model contains more than just api_key, that's the only bit that is ignored, so the union wouldn't work. Do you still think it makes sense?

girarda replied:
can we fail if the API key is set?

flash1293 replied:
Sounds like a good approach

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:
```diff
@@ -1010,6 +1010,81 @@ definitions:
+        type: array
+        items:
+          type: string
+      expiration_time:
```
nit: can we call this `expiration_duration`, `expires_in`, or something else that makes it clear this is a duration and not a timestamp

girarda commented on airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py:
```diff
@@ -155,3 +168,343 @@ def cache_filename(self) -> str:
+        next_page_token: Optional[Mapping[str, Any]],
+        requester_method,
+        auth_options_method,
+        extra_options: Optional[Mapping[str, Any]] = None,
```
it looks like extra_options can also be a string if this method is called from [_request_body_data](https://github.com/airbytehq/airbyte/pull/28050/files#diff-17485be091db6243dff1b4ef8d77a71b199fd37a09be7cbaa3ebaaea4228ed29R315). Is that expected?

flash1293 replied:
Yes, I just forgot to adjust the type. If it's a string it will make sure there is only one and throw otherwise.

girarda reviewed (commented):
looks good. just a few small comments.

Do you plan on updating the docs in this PR or in a follow up?

flash1293 commented:
> Do you plan on updating the docs in this PR or in a follow up?

Which docs are you referring to? The connector builder docs will be updated once there's a UI for this.

I addressed your points except for the model_to_component factory one, not sure how to structure things to make them much clearer than they are right now.

flash1293 commented:
@girarda Added the check for empty api_token

octavia-squidington-iii commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/bcb901a333943b694a5ba0241620d68f285b18df/airbyte-integrations/connectors/source-apify-dataset/icon.svg" width="40" height="40"> source-apify-dataset test report (commit [`bcb901a333`](https://github.com/airbytehq/airbyte/pull/28050/commits/bcb901a333943b694a5ba0241620d68f285b18df)) - ❌

⏲️  Total pipeline duration: 03mn44s 

| Step                                                                        | Result   |
|:----------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-apify-dataset/metadata.yaml | ✅        |
| Connector version semver check                                              | ✅        |
| Connector version increment check                                           | ❌        |
| QA checks                                                                   | ✅        |
| Code format checks                                                          | ✅        |
| Connector package install                                                   | ✅        |
| Build source-apify-dataset docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                  | ✅        |
| Acceptance tests                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_session-token-authenticator/1689675909/bcb901a333943b694a5ba0241620d68f285b18df/source-apify-dataset/0.2.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-apify-dataset test
```

octavia-squidington-iii commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/bcb901a333943b694a5ba0241620d68f285b18df/airbyte-integrations/connectors/destination-bigquery/icon.svg" width="40" height="40"> destination-bigquery test report (commit [`bcb901a333`](https://github.com/airbytehq/airbyte/pull/28050/commits/bcb901a333943b694a5ba0241620d68f285b18df)) - ❌

⏲️  Total pipeline duration: 42mn19s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_session-token-authenticator/1689675909/bcb901a333943b694a5ba0241620d68f285b18df/destination-bigquery/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

octavia-squidington-iii commented:
## <img src="https://raw.githubusercontent.com/airbytehq/airbyte/bcb901a333943b694a5ba0241620d68f285b18df/airbyte-integrations/connectors/destination-bigquery-denormalized/icon.svg" width="40" height="40"> destination-bigquery-denormalized test report (commit [`bcb901a333`](https://github.com/airbytehq/airbyte/pull/28050/commits/bcb901a333943b694a5ba0241620d68f285b18df)) - ✅

⏲️  Total pipeline duration: 21mn03s 

| Step                                                                                         | Result   |
|:---------------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery-denormalized/metadata.yaml     | ✅        |
| Connector version semver check                                                               | ✅        |
| QA checks                                                                                    | ✅        |
| Build connector tar                                                                          | ✅        |
| Build destination-bigquery-denormalized docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery-denormalized:integrationTest | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/flash1293_session-token-authenticator/1689675909/bcb901a333943b694a5ba0241620d68f285b18df/destination-bigquery-denormalized/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery-denormalized test
```

Marcaccio-Mutt commented:
Hello!

I have a question about the parameter `expiration_duration`. Is it working? Although it has a value, the token is updated in each request in my case :thinking:

Marcaccio-Mutt commented:
Hello again!

I would like to know if you have some advice about my previous question... I will be so grateful if you could help me with this
